### ISSUE TITLE
Improve handling of Fortran case insensitivity

### ISSUE THREAD

**Issue by RaoulHC** [universal-ctags/ctags#3668]
Fortran is a case insensitive language and as a result it would be great if ctags took this into account.

For example the following code defines a subroutine as `FOO` but can happily be called by `foo`, and similarly uses `myvar` in multiple ways, but the tags file generated from it is not as useful as I'd like.

```Fortran
      subroutine FOO()
      integer myVar
      myvar = 3
      print *, "myVar is ", MyVar
      end subroutine

      program my_program
      call foo()
      end program my_program
```
```
$ ./ctags --kinds-Fortran=+L -o - test.f
FOo     test.f  /^      subroutine FOo(/;"      s
myVar   test.f  /^      integer myVar$/;"       L       subroutine:FOo  file:
my_program      test.f  /^      program my_program$/;"  p
```

Additionally Fortran symbols are appended with an underscore by most compilers, meaning that if in a multi language codebase I want to go to the definition of a Fortran function called in c as `foo_` I similarly wont find it.

I think it makes sense for ctags to lowercase symbols, and potentially make sense for it to append an underscore to subroutine and function names. However because the latter may not play well with other tooling (my usecase is with opengrok), it might make the most sense to generate two tags, with and without the underscore, probably gated behind a flag.

Looks like it should be easy enough to implement this in the Fortran parser but as I'm not all that familiar with ctags I'd appreciate some input to this problem before raising a PR.

**Comment by masatake** [comment]
Should ctags generate the following tags for the former reqeust?

```
FOO     test.f  /^      subroutine FOO(/;"      s
FOo     test.f  /^      subroutine FOO(/;"      s
FoO     test.f  /^      subroutine FOO(/;"      s
FOo     test.f  /^      subroutine FOO(/;"      s
Foo     test.f  /^      subroutine FOO(/;"      s
fOO     test.f  /^      subroutine FOO(/;"      s
fOo     test.f  /^      subroutine FOO(/;"      s
foO     test.f  /^      subroutine FOO(/;"      s
foo     test.f  /^      subroutine FOO(/;"      s
```

**Comment by masatake** [comment]
If ctags records `foo` in a tags file, searching `FOO` with a tool consuming the tags file may not work, and it is incompatible destructive behavior for tools having been working with ctags.
Unacceptable. These things should be solved with consumers of tag files. I'm negative.

`foo_` is a bit interesting. How do we call the id like `foo_`.  ELF name? Link name?

**Comment by masatake** [comment]
https://stackoverflow.com/questions/1569887/is-it-possible-to-perform-a-case-sensitive-search-in-opengrok

It seems that case sensitivity can be handled in OpenGrok side.

**Comment by RaoulHC** [comment]
> If ctags records `foo` in a tags file, searching `FOO` with a tool consuming the tags file may not work, and it is incompatible destructive behavior for tools having been working with ctags. Unacceptable. These things should be solved with consumers of tag files. I'm negative.

Perhaps this should be gated by a flag? As it stands if a function is defined as `MyFunction`, but is called as `myfunction`, if a user searches for the latter they will not find it, which is not ideal behaviour.
 
> `foo_` is a bit interesting. How do we call the id like `foo_`. ELF name? Link name?

I think link name makes sense.

> It seems that case sensitivity can be handled in OpenGrok side.

Note there that the full search is case insensitive, but the search for definition is case sensitive. This is what I think would need to be changed in opengrok for Fortran specifically.

**Comment by RaoulHC** [comment]
> Could you tell me a documentation or specification about the rule synthesizing the name like foo_?

It's not part of the standard, but is the default behaviour for most compilers, for example [gfortran](https://gcc.gnu.org/onlinedocs/gfortran/Code-Gen-Options.html), [Sun Studio](https://docs.oracle.com/cd/E19957-01/805-4940/z400091044a7/index.html). By default, XLF does not do it, but it can be turned on by [-qextname](https://www.ibm.com/docs/en/openxl-fortran-aix/17.1.1?topic=options-qextname).

> Which kind may have such a link name?

I'll double check, but off the top of my head I believe the following symbols have underscore appended: common, entry, function, interface, and subroutine. Unsure about some of the module features, and block data, but stuff like variable and types don't have any linkage significance.

**Comment by masatake** [comment]
As far as reading the man page for GNU Fortran you kindly introduced, the compiler appends two underscores to a link name whose original name includes an underscore.

e.g.
`my_var__` is for `my_var`.
`foo_` is for `foo`.

Do you think ctags should do the same?

**Comment by RaoulHC** [comment]
Hmmm I think the GFortran docs are actually slightly misleading, it describes what's happening with the non-default `-fsecond-underscore` option under the `-fno-underscore` option.

Testing it on the following example:
```fortran
      program main
      call subr()
      call my_subr()
      end program main
```
Compiling it with the following options and inspecting the symbols in the resulting object file I get:
* `-funderscoring` (the default): `subr_`, `my_subr_`
* `-fno-underscoring`: `subr`, `my_subr`
* `-fsecond-underscore`: `subr_`, `my_subr__`

I'll raise an issue with GCC folks to improve that, but I think adding a single underscore regardless of how many underscores there are in the name makes sense.

**Comment by masatake** [comment]
@RaoulHC Thank you. Based on your advice I made a pull request. See #3671.

**Comment by RaoulHC** [comment]
Excellent, looks like some tests need updating, but otherwise that looks like that'll really help our opengrok instance find Fortran definitions.

Thanks so much for this!

**Comment by masatake** [comment]
> Excellent, looks like some tests need updating, ...

Please, give comments on the pull request. I will not merge till I get your ack.
